flink/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.actions;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import org.apache.flink.api.java.DataSet;
+import org.apache.flink.api.java.ExecutionEnvironment;
+import org.apache.iceberg.CombinedScanTask;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.flink.source.RowDataRewriter;
+import org.apache.iceberg.util.StructLikeWrapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RewriteDataFilesAction extends
+    RewriteDataFilesActionBase<RewriteDataFilesAction, RewriteDataFilesActionResult> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RewriteDataFilesAction.class);
+
+  private ExecutionEnvironment env = null;
+  private boolean caseSensitive;
+
+  public RewriteDataFilesAction(ExecutionEnvironment env, Table table) {
+    super(table);
+    this.env = env;
+    caseSensitive = Boolean.parseBoolean(table.properties().getOrDefault("case-sensitive", "false"));
+  }
+
+
+  @Override
+  public RewriteDataFilesActionResult execute() {
+    Map<StructLikeWrapper, Collection<FileScanTask>> filteredGroupedTasks = getFilteredGroupedTasks();
+
+    // Nothing to rewrite if there's only one DataFile in each partition.
+    if (filteredGroupedTasks.isEmpty()) {
+      return RewriteDataFilesActionResult.empty();
+    }
+
+    List<DataFile> currentDataFiles = getCurrentDataFiles(filteredGroupedTasks);
+
+    List<CombinedScanTask> combinedScanTasks = getCombinedScanTasks(filteredGroupedTasks);
+
+    DataSet<CombinedScanTask> dataSet = env.fromCollection(combinedScanTasks);
+    env.setParallelism(combinedScanTasks.size());
+
+    RowDataRewriter rowDataRewriter =
+        new RowDataRewriter(getTable(), getTable().spec(), caseSensitive, getTable().io(), getEncryptionManager());
+
+    List<DataFile> addedDataFiles = null;
+    try {
+      addedDataFiles = rowDataRewriter.rewriteDataForTasks(dataSet);
+    } catch (Exception e) {
+      LOG.error("rewrite data files error", e);
```

Comment:
I think here we'd better to throw the exception to upper layer ?  How about throwing the `UncheckedIOException(e)` ?

Reply:
This exception comes from the DataSet#collect method. I looked at this method again. It throws a RuntimeException. I think we also throw a RuntimeException here. What do you think?

Reply:
Both `RuntimeException`  and `UncheckedIOException` works fine for me. Actually,  `UncheckedIOException` is also a `RuntimeException`.